esp/esp/users/models/__init__.py

```diff
@@ -2364,18 +2395,34 @@ def user_has_perm(cls, user, name, program=None, when=None, program_is_none_impl
         """
         if user.isAdministrator(program=program):
             return True
-        if name in cls.deadline_types:
-            program_is_none_implies_all = False
-        perms=[name]
-        for k,v in cls.implications.items():
-            if name in v: perms.append(k)
 
         quser = Q(user=user) | Q(user=None, role__in=user.groups.all())
-        qprogram = Q(program=program)
-        if program_is_none_implies_all:
-            qprogram |= Q(program=None)
-        initial_qset = cls.objects.filter(quser & qprogram).filter(permission_type__in=perms)
-        return initial_qset.filter(cls.is_valid_qobject(when=when)).exists()
+        return cls.q_permissions_on_program(quser, name, program, when,
+                program_is_none_implies_all).exists()
+
+    @classmethod
+    def list_roles_with_perm(cls, name, program):
+        """Given a permission type on a program, list roles that would give the
+        permission on the program.
+
+        :param name:
+            The unique identifier of the permission identifier to check for.
+            Must be in PERMISSION_CHOICES_FLAT.
+        :type name:
+            `str`
+        :param program:
+            Check for permission for `name` on this program.
+        :type program:
+            `Program`
+        :return:
+            List of role names that would give the specified permission.
+        :rtype:
+            `list` of `str`
+        """
+
+        qrole = Q(user=None, role__isnull=False)
+        return [permission.role.name for permission in
```

Comment:
Probably slightly better to do `.values_list('role__name')` to avoid a second db fetch for the related lookup (and send less data back from the db to boot).